File: graphs/ai_trader/tools/compile.py
"""Compile tools for QuantConnect."""

import asyncio
import json

from langchain.tools import tool, ToolRuntime
from langgraph.graph.ui import push_ui_message

from ..context import Context
from ..qc_api import qc_request


@tool
async def create_compile(runtime: ToolRuntime[Context]) -> str:
    """
    Compile the current project on QuantConnect.
    Returns the compile ID needed for backtests and optimizations.
    """
    try:
        qc_project_id = runtime.context.get("qc_project_id")
        user_id = runtime.context.get("user_id")

        if not qc_project_id:
            return json.dumps({"error": True, "message": "No project context."})

        result = await qc_request(
            "/compile/create",
            {"projectId": qc_project_id},
            user_id=user_id,
        )

        compile_id = result.get("compileId")
        state = result.get("state", "Unknown")

        # Emit compiling UI
        push_ui_message("compile-status", {
            "compileId": compile_id,
            "state": "InQueue",
            "status": "Compiling...",
        }, message={"id": runtime.tool_call_id})

        # Wait for compilation to complete
        max_wait = 60
        waited = 0
        while state == "InQueue" and waited < max_wait:
            await asyncio.sleep(2)
            waited += 2
            status = await qc_request(
                "/compile/read",
                {"projectId": qc_project_id, "compileId": compile_id},
                user_id=user_id,
            )
            state = status.get("state", "Unknown")

        if state == "BuildSuccess":
            # Emit success UI
            push_ui_message("compile-status", {
                "compileId": compile_id,
                "state": "BuildSuccess",
                "status": "Compilation successful",
                "success": True,
            }, message={"id": runtime.tool_call_id})
            
            return json.dumps(
                {
                    "success": True,
                    "compile_id": compile_id,
                    "state": state,
                    "message": "Compilation successful. Ready for backtest or optimization.",
                }
            )
        elif state == "BuildError":
            logs = result.get("logs", [])
            
            # Emit error UI
            push_ui_message("compile-status", {
                "compileId": compile_id,
                "state": "BuildError",
                "status": "Compilation failed",
                "success": False,
                "errors": logs[:5] if logs else [],
            }, message={"id": runtime.tool_call_id})
            
            return json.dumps(
                {
                    "error": True,
                    "compile_id": compile_id,
                    "state": state,
                    "logs": logs[:20] if logs else [],
                    "message": "Compilation failed.",
                }
            )
        else:
            # Emit status UI
            push_ui_message("compile-status", {
                "compileId": compile_id,
                "state": state,
                "status": f"Compile state: {state}",
            }, message={"id": runtime.tool_call_id})
            
            return json.dumps(
                {
                    "success": True,
                    "compile_id": compile_id,
                    "state": state,
                    "message": f"Compile created with state: {state}",
                }
            )

    except Exception as e:
        return json.dumps({"error": True, "message": f"Failed to compile: {e!s}"})


@tool
async def read_compile(
    compile_id: str,
    runtime: ToolRuntime[Context],
) -> str:
    """
    Read compile status and logs.

    Args:
        compile_id: The compile ID to check
    """
    try:
        qc_project_id = runtime.context.get("qc_project_id")
        user_id = runtime.context.get("user_id")

        if not qc_project_id:
            return json.dumps({"error": True, "message": "No project context."})

        result = await qc_request(
            "/compile/read",
            {"projectId": qc_project_id, "compileId": compile_id},
            user_id=user_id,
        )

        state = result.get("state", "Unknown")
        logs = result.get("logs", [])

        # Emit compile status UI
        push_ui_message("compile-status", {
            "compileId": compile_id,
            "state": state,
            "status": "Success" if state == "BuildSuccess" else "Failed" if state == "BuildError" else state,
            "success": state == "BuildSuccess",
            "errors": logs[:5] if state == "BuildError" and logs else [],
        }, message={"id": runtime.tool_call_id})

        return json.dumps(
            {
                "compile_id": compile_id,
                "state": state,
                "logs": logs[:20] if logs else [],
            },
            indent=2,
        )

    except Exception as e:
        return json.dumps({"error": True, "message": f"Failed to read compile: {e!s}"})


# Export all tools
TOOLS = [create_compile, read_compile]

File: packages/lean-api/src/routes/compile.ts
/**
 * Compile Routes - QC API Compatible
 * POST /api/v2/compile/create
 */

import { Router, type IRouter } from 'express';
import { query, queryOne } from '../services/database.js';
import { v4 as uuidv4 } from 'uuid';
import { logError, formatErrorForResponse, getErrorStatusCode } from '../utils/errors.js';
import type {
  QCCompileResponse,
  CompileCreateRequest,
  ProjectFile,
} from '../types/index.js';

const router: IRouter = Router();

/**
 * Look up project by QC project ID and verify ownership
 * Uses the main 'projects' table which has qc_project_id
 * Returns internal project id if found, null otherwise
 */
async function getProjectByQcId(qcProjectId: number, userId: string): Promise<number | null> {
  // Look up by qc_project_id in projects table
  const project = await queryOne<{ id: number }>(
    'SELECT id FROM projects WHERE qc_project_id = $1 AND user_id = $2',
    [String(qcProjectId), userId]
  );
  if (project) {
    return project.id;
  }

  // Fallback: maybe it's the internal project id directly
  const directProject = await queryOne<{ id: number }>(
    'SELECT id FROM projects WHERE id = $1 AND user_id = $2',
    [qcProjectId, userId]
  );
  return directProject?.id || null;
}

/**
 * Basic Python syntax validation
 *
 * IMPORTANT: This is intentionally minimal to avoid false positives.
 * Python syntax is complex (multi-line statements, f-strings, etc.)
 * and naive checks cause valid QC code to fail.
 *
 * We only check for things we can reliably detect:
 * - Required LEAN imports
 * - QCAlgorithm inheritance
 * - Initialize method presence
 */
function validatePythonSyntax(code: string): { valid: boolean; errors: string[] } {
  const errors: string[] = [];

  // Check for required LEAN imports
  if (!code.includes('from AlgorithmImports import') && !code.includes('import AlgorithmImports')) {
    errors.push('Missing required import: from AlgorithmImports import *');
  }

  // Check for QCAlgorithm class
  if (!code.includes('QCAlgorithm')) {
    errors.push('Algorithm must inherit from QCAlgorithm');
  }

  // Check for initialize method (case-insensitive check)
  if (!code.includes('def initialize(self)') && !code.includes('def Initialize(self)')) {
    errors.push('Algorithm must have an initialize(self) method');
  }

  return {
    valid: errors.length === 0,
    errors,
  };
}

/**
 * POST /compile/create - Compile/validate a project
 */
router.post('/create', async (req, res) => {
  const context = { endpoint: 'compile/create', userId: req.userId, body: req.body };

  try {
    const { projectId } = req.body as CompileCreateRequest;
    const userId = req.userId;

    if (!projectId) {
      return res.status(400).json({
        success: false,
        compileId: '',
        projectId: 0,
        state: 'BuildError',
        parameters: [],
        signature: '',
        signatureOrder: [],
        logs: ['projectId is required'],
        errors: ['projectId is required'],
      });
    }

    if (typeof projectId !== 'number' || projectId < 1) {
      return res.status(400).json({
        success: false,
        compileId: '',
        projectId: projectId || 0,
        state: 'BuildError',
        parameters: [],
        signature: '',
        signatureOrder: [],
        logs: ['projectId must be a positive integer'],
        errors: ['projectId must be a positive integer'],
      });
    }

    // Look up internal project id from QC project id
    const internalProjectId = await getProjectByQcId(projectId, userId);
    if (!internalProjectId) {
      return res.status(404).json({
        success: false,
        compileId: '',
        projectId,
        state: 'BuildError',
        parameters: [],
        signature: '',
        signatureOrder: [],
        logs: ['Project not found or access denied'],
        errors: ['Project not found or access denied'],
      });
    }

    const files = await query<ProjectFile>(
      'SELECT * FROM project_files WHERE project_id = $1',
      [internalProjectId]
    );

    if (files.length === 0) {
      return res.status(400).json({
        success: false,
        compileId: '',
        projectId,
        state: 'BuildError',
        parameters: [],
        signature: '',
        signatureOrder: [],
        logs: ['No files found in project'],
        errors: ['No files found in project'],
      });
    }

    const mainFile = files.find(f => f.isMain || f.name.toLowerCase() === 'main.py');
    if (!mainFile) {
      return res.status(400).json({
        success: false,
        compileId: '',
        projectId,
        state: 'BuildError',
        parameters: [],
        signature: '',
        signatureOrder: [],
        logs: ['main.py not found'],
        errors: ['main.py not found'],
      });
    }

    const allLogs: string[] = [];
    let hasErrors = false;

    for (const file of files) {
      if (file.name.endsWith('.py')) {
        const validation = validatePythonSyntax(file.content);
        if (!validation.valid) {
          hasErrors = true;
          allLogs.push(`Errors in ${file.name}:`);
          allLogs.push(...validation.errors.map(e => `  ${e}`));
        }
      }
    }

    const compileId = uuidv4();

    if (hasErrors) {
      return res.json({
        success: false,
        compileId,
        projectId,
        state: 'BuildError',
        parameters: [],
        signature: '',
        signatureOrder: [],
        logs: allLogs,
        errors: allLogs,
      });
    }

    res.json({
      success: true,
      compileId,
      projectId,
      state: 'BuildSuccess',
      parameters: [],
      signature: compileId,
      signatureOrder: files.map(f => f.name),
      logs: ['Syntax validation passed', 'Ready to run backtest'],
      errors: [],
    });
  } catch (error) {
    logError('compile/create', error, context);
    const statusCode = getErrorStatusCode(error);
    res.status(statusCode).json({
      success: false,
      compileId: '',
      projectId: req.body?.projectId || 0,
      state: 'BuildError',
      parameters: [],
      signature: '',
      signatureOrder: [],
      logs: [formatErrorForResponse(error)],
      errors: [formatErrorForResponse(error)],
    });
  }
});

/**
 * POST /compile/read - Read compile status
 *
 * QC API Compatible: Returns compilation state for a given compile ID.
 * For self-hosted LEAN, compilation is synchronous so this always returns
 * the same state as /compile/create.
 */
router.post('/read', async (req, res) => {
  const context = { endpoint: 'compile/read', userId: req.userId, body: req.body };

  try {
    const { projectId, compileId } = req.body as { projectId: number; compileId: string };
    const userId = req.userId;

    if (!projectId) {
      return res.status(400).json({
        success: false,
        compileId: compileId || '',
        projectId: 0,
        state: 'BuildError',
        parameters: [],
        signature: '',
        signatureOrder: [],
        logs: ['projectId is required'],
        errors: ['projectId is required'],
      });
    }

    if (!compileId) {
      return res.status(400).json({
        success: false,
        compileId: '',
        projectId,
        state: 'BuildError',
        parameters: [],
        signature: '',
        signatureOrder: [],
        logs: ['compileId is required'],
        errors: ['compileId is required'],
      });
    }

    // Look up internal project id from QC project id
    const internalProjectId = await getProjectByQcId(projectId, userId);
    if (!internalProjectId) {
      return res.status(404).json({
        success: false,
        compileId,
        projectId,
        state: 'BuildError',
        parameters: [],
        signature: '',
        signatureOrder: [],
        logs: ['Project not found or access denied'],
        errors: ['Project not found or access denied'],
      });
    }

    // For self-hosted LEAN, compilation is synchronous (done in /compile/create)
    // So /compile/read just re-validates the current files state

    const files = await query<ProjectFile>(
      'SELECT * FROM project_files WHERE project_id = $1',
      [internalProjectId]
    );

    if (files.length === 0) {
      return res.status(400).json({
        success: false,
        compileId,
        projectId,
        state: 'BuildError',
        parameters: [],
        signature: '',
        signatureOrder: [],
        logs: ['No files found in project'],
        errors: ['No files found in project'],
      });
    }

    const mainFile = files.find(f => f.isMain || f.name.toLowerCase() === 'main.py');
    if (!mainFile) {
      return res.status(400).json({
        success: false,
        compileId,
        projectId,
        state: 'BuildError',
        parameters: [],
        signature: '',
        signatureOrder: [],
        logs: ['main.py not found'],
        errors: ['main.py not found'],
      });
    }

    // Re-validate the files
    const allLogs: string[] = [];
    let hasErrors = false;

    for (const file of files) {
      if (file.name.endsWith('.py')) {
        const validation = validatePythonSyntax(file.content);
        if (!validation.valid) {
          hasErrors = true;
          allLogs.push(`Errors in ${file.name}:`);
          allLogs.push(...validation.errors.map(e => `  ${e}`));
        }
      }
    }

    if (hasErrors) {
      return res.json({
        success: false,
        compileId,
        projectId,
        state: 'BuildError',
        parameters: [],
        signature: '',
        signatureOrder: [],
        logs: allLogs,
        errors: allLogs,
      });
    }

    // Return success - compilation is complete for self-hosted LEAN
    res.json({
      success: true,
      compileId,
      projectId,
      state: 'BuildSuccess',
      parameters: [],
      signature: compileId,
      signatureOrder: files.map(f => f.name),
      logs: ['Syntax validation passed', 'Ready to run backtest'],
      errors: [],
    });
  } catch (error) {
    logError('compile/read', error, context);
    const statusCode = getErrorStatusCode(error);
    res.status(statusCode).json({
      success: false,
      compileId: req.body?.compileId || '',
      projectId: req.body?.projectId || 0,
      state: 'BuildError',
      parameters: [],
      signature: '',
      signatureOrder: [],
      logs: [formatErrorForResponse(error)],
      errors: [formatErrorForResponse(error)],
    });
  }
});

export default router;
